### PR TITLE
Add elm workshop article.

### PR DESCRIPTION
**Pull request by Natim**
Looking forward for your review.

https://5c9a6c58a6a64a00072a7756--jolly-hugle-47d008.netlify.com/getting-started-with-elm-0.19

### CODE REVIEW

File: content/2019.04.getting-started-with-elm.rst [hunk 1]
```diff
@@ -0,0 +1,1100 @@
...
+by inconsistent typing. Being functional and pure, Elm let you write code
+that is completely decoupled, which makes your code more reusable and
+easier to refactor.
+
+
+How is Elm different?
+=====================
+
+Most of the time, using Elm, the two lines it takes to load your app
+from the HTML file is all the JavaScript you will write, as your Elm
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
from the HTML file is all the JavaScript you need. Your Elm
```

File: content/2019.04.getting-started-with-elm.rst [hunk 2]
```diff
@@ -0,0 +1,1100 @@
...
+code will be compiled to JavaScript by the Elm compiler.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
code is compiled to JavaScript by the Elm compiler.
```

Remove future tense

File: content/2019.04.getting-started-with-elm.rst [hunk 3]
```diff
@@ -0,0 +1,1100 @@
...
+
+Most mistakes you make will be caught at compilation time, which
+means that you will barely ever encounter exceptions at runtime in the
+browser. Also that means less unit tests and defensive programming to
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
browser. Also, that means fewer unit tests and defensive programming to
```

File: content/2019.04.getting-started-with-elm.rst [hunk 4]
```diff
@@ -0,0 +1,1100 @@
...
+        name
+        tan
+        Cmd.none
+        Sub.none
+
+    Detected errors in 1 module.
+
+The tooling is great. elm-live_ automatically updates the browser
+code to reflect your code update in realtime. It comes with a debugger
+that will show you the list of events and state of your application at
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
that shows you the list of events and state of your application at
```

File: content/2019.04.getting-started-with-elm.rst [hunk 5]
```diff
@@ -0,0 +1,1100 @@
...
+a given point in time — you can even go back in time and replay previous
+events. elm-format_ will also format your code automatically to avoid
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
events. elm-format_ also formats your code automatically avoiding arguments
```

File: content/2019.04.getting-started-with-elm.rst [hunk 6]
```diff
@@ -0,0 +1,1100 @@
...
+fighting over coding styles.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
 over coding styles.
```

File: content/2019.04.getting-started-with-elm.rst [hunk 7]
```diff
@@ -0,0 +1,1100 @@
...
+    my project? How do I see it in the browser? How will my code grow? Do I need
+    more directories? What about tests? Etc.
+
+    Check out <https://elm-lang.org/0.19.0/init> for all the answers!
+
+    Knowing all that, would you like me to create an elm.json file now? [Y/n]: 
+    Okay, I created it. Now read that link!
+
+You now have an ``elm.json`` file in your project as well as an empty
+``src/`` directory where you will put your source files.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
``src/`` directory where you put your source files.
```

File: content/2019.04.getting-started-with-elm.rst [hunk 8]
```diff
@@ -0,0 +1,1100 @@
...
+    └── src
+
+    1 directory, 1 file
+
+
+Creating your first file
+========================
+
+To get started you can simply create a new file named ``src/Main.elm``
+with the following content:
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
:
```

File: content/2019.04.getting-started-with-elm.rst [hunk 9]
```diff
@@ -0,0 +1,1100 @@
...
+
+.. code-block:: elm
+
+    import Html
+    main = Html.text "Hello world"
+
+Elm benefits from a full featured module system, with a broad
+ecosystem of external packages available. ``Html`` is part of the core.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
ecosystem of external packages available. ``Html`` ships with Elm, it is part of the core libs.
```
maybe ~ ?

File: content/2019.04.getting-started-with-elm.rst [hunk 10]
```diff
@@ -0,0 +1,1100 @@
...
+                    ]
+                ]
+            ]
+
+
+Adding some state
+=================
+
+Now that you know how to render your page in HTML, let's see how to
+write a program that handles events.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]: DOM events ?

File: content/2019.04.getting-started-with-elm.rst [hunk 11]
```diff
@@ -0,0 +1,1100 @@
...
+
+The way Elm handles that is by having:
+
+- a ``Model`` that keep the state of the app,
+- an ``update`` function that will handle all the app events and update
+  the model state accordingly
+- a view function that will return the Virtual DOM that matches the state
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
- a view function that returns the Virtual DOM matching the state of the app
```

File: content/2019.04.getting-started-with-elm.rst [hunk 12]
```diff
@@ -0,0 +1,1100 @@
...
+Adding some state
+=================
+
+Now that you know how to render your page in HTML, let's see how to
+write a program that handles events.
+
+The way Elm handles that is by having:
+
+- a ``Model`` that keep the state of the app,
+- an ``update`` function that will handle all the app events and update
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
- an ``update`` function that handles all the app events and updates
```

File: content/2019.04.getting-started-with-elm.rst [hunk 13]
```diff
@@ -0,0 +1,1100 @@
...
+For those who know Redux_, it has been heavily inspired by
+Elm. Basically ``update`` is a reducer.
+
+.. _Redux: http://redux.js.org/
+
+The events and their parameters are defined in a ``Msg``.
+
+In order to create our application that handle states, we can use the
+``sandbox`` from the Browser package:
+
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]: Small sentence introducing about sandbox and a link to `see more` may be a good idea.

**Comment by Natim** [comment]
- [ ] Add an example of Redux vs Elm
- [ ] Explain types with the common `type Bool = True | False` example
- [ ] Explain Browser.sandbox Browser.element, Browser.document

File: content/2019.04.getting-started-with-elm.rst [hunk 14]
```diff
@@ -0,0 +1,1100 @@
...
+Conclusion
+==========
+
+That's about it. Now that you understand how the event update mechanism works and how
+you can define functions, you know more than you think about Elm.
+
+When in doubt, the package documentation is really useful: https://package.elm-lang.org/
+
+I hope you try it on your next project and enjoy Elm as much as we do.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
I hope you give Elm a shot on your next project and enjoy Elm as much as we do @Chefclub.
```

File: content/2019.04.getting-started-with-elm.rst [hunk 15]
```diff
@@ -0,0 +1,1100 @@
...
+
+Wait a minute, That's it? Do I really know everything? But you didn't
+tell me how I was supposed to handle HTTP requests yet!
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]: 🤣

File: content/2019.04.getting-started-with-elm.rst [hunk 16]
```diff
@@ -0,0 +1,1100 @@
...
+
+
+Handling HTTP requests
+======================
+
+Fair enough, I remember I asked exactly this question when I was
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
Fair enough, I remember asking exactly this question when I was
```

File: content/2019.04.getting-started-with-elm.rst [hunk 17]
```diff
@@ -0,0 +1,1100 @@
...
+introduced to Elm_.
+
+Let's use the `photos collection`_ of `JSON Placeholder`_ to get a list of JSON objects.
+
+.. _`photos collection`: https://jsonplaceholder.typicode.com/photos
+.. _`JSON Placeholder`: https://jsonplaceholder.typicode.com/
+
+In order to do so we will use the `elm/http`_ library.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
In order to do so we use the `elm/http`_ library.
```

File: content/2019.04.getting-started-with-elm.rst [hunk 18]
```diff
@@ -0,0 +1,1100 @@
...
+         (Decode.field "title" Decode.string)
+         (Decode.field "url" Decode.string)
+         (Decode.field "thumbnailUrl" Decode.string)
+
+
+At this stage you might be wondering what is this ``map4`` thing and
+why on Earth we would use a function with the number of field that we
+want to decode.
+
+I am glad you are asking yourself this.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
I am glad you asked ;)
```

File: content/2019.04.getting-started-with-elm.rst [hunk 19]
```diff
@@ -0,0 +1,1100 @@
...
+
+To go back a little bit, there are two ways of creating a record:
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
Let's rewind a little bit, there are two ways of creating a record:
```

File: content/2019.04.getting-started-with-elm.rst [hunk 20]
```diff
@@ -0,0 +1,1100 @@
...
+           , url = "https://profile.nytimes.com/accounts/1.png"
+           , thumbnailUrl = "https://profile.nytimes.com/accounts/thumbs/1.png"
+           }
+
+Decoders are using the constructor way to create records.
+
+Decoding JSON values into records using ``Decode.map#``
+-------------------------------------------------------
+
+``Decode.map#`` will decode each fields and then build a record using the
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
``Decode.map#`` decodes each fields and then build a record using the
```

File: content/2019.04.getting-started-with-elm.rst [hunk 21]
```diff
@@ -0,0 +1,1100 @@
...
+the decoded fields is important and should be the same as the type
+alias definition.
+
+
+Decoding JSON values into records using ``Decode.succeed``
+----------------------------------------------------------
+
+We can also create a record and use ``Decode.succeed`` to mark it as
+a decoded value. That's what `NoRedInk/elm-json-decode-pipeline`_ is
+providing.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]: What does elm-json-decode-pipeline provides ?

File: content/2019.04.getting-started-with-elm.rst [hunk 22]
```diff
@@ -0,0 +1,1100 @@
...
+
+Note that in that case field order is also important, this will
+compose a decoder that in the end returns an object and in between
+return partial decoding functions.
+
+
+Handling the response
+---------------------
+
+Once the response body has been decoded, elm will send a message with
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
Once the response body has been decoded, elm sends a message with
```

Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
Once the response body has been decoded, elm sends a message with
```

File: content/2019.04.getting-started-with-elm.rst [hunk 23]
```diff
@@ -0,0 +1,1100 @@
...
+a result that will be handled by the update function.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
a result that is handled by the update function.
```

File: content/2019.04.getting-started-with-elm.rst [hunk 24]
```diff
@@ -0,0 +1,1100 @@
...
+    
+            GotItems (Err err) ->
+                ( { model | error = Just <| errorToString err, photos = [] }, Cmd.none )
+
+
+Showing the list of pictures
+============================
+
+Once we are able to fetch our list of photos, we might want to display
+it.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
those pretty pictures.
```

File: content/2019.04.getting-started-with-elm.rst [hunk 25]
```diff
@@ -0,0 +1,1100 @@
...
+
+I invite you to have a look at the Ellie with the fully functionnal
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]: Whats Ellie ?

File: content/2019.04.getting-started-with-elm.rst [hunk 26]
```diff
@@ -0,0 +1,1100 @@
...
+``Html Msg``. We can then use this result as a list of children to a ``div []``
+element.
+
+.. _`List.map`: https://package.elm-lang.org/packages/elm/core/latest/List#map
+
+
+Conclusion
+==========
+
+What's next ? Starting from here, you can grow your widget. At some
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
What's next? Starting from here, you can grow your widget. At some
```

File: content/2019.04.getting-started-with-elm.rst [hunk 27]
```diff
@@ -0,0 +1,1100 @@
...
+point you might want to create a Single Page App (SPA) and handle URL
+with multiple pages.
+
+That's where the Elm Architecture start to be really useful. I would
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
That's where the Elm Architecture really starts to shine. I would
```

File: content/2019.04.getting-started-with-elm.rst [hunk 28]
```diff
@@ -0,0 +1,1100 @@
...
+recommend you to have a look at `elm-kitchen`_ which will help you to get
+started with it.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
started with the scaffolding.
```

?

File: content/2019.04.getting-started-with-elm.rst [hunk 29]
```diff
@@ -0,0 +1,1100 @@
...
+Elm types
+---------
+
+In Elm everything has a type.
+
+- ``"hello"`` is a ``String``
+- ``4`` is a ``number``
+- ``4.2`` is a ``Float``
+
+Elm itself define the usual types, however our business logic
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
Elm itself defines the usual types, however, our business logic
```

File: content/2019.04.getting-started-with-elm.rst [hunk 30]
```diff
@@ -0,0 +1,1100 @@
...
+Let's think about a user, it can be ``Active`` or ``Inactive``.
+
+In other languages we would use an Enum, in Elm_ we can use a type.
+
+.. code-block:: elm
+
+    type Status = Active | Inactive
+
+You might want to use a boolean for this specific case, however using
+a type here will make your code more readable.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
a type here does make your code more readable.
```

File: content/2019.04.getting-started-with-elm.rst [hunk 31]
```diff
@@ -0,0 +1,1100 @@
...
+
+The benefit of use a type is that Elm will be able to validate that
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
The benefit of use a type is that Elm is able to validate that
```

File: content/2019.04.getting-started-with-elm.rst [hunk 32]
```diff
@@ -0,0 +1,1100 @@
...
+
+    type alias User =
+         { username : String, status : Status }
+
+    getUserName :  User -> String
+    getUserName user =
+        user.username
+
+Because ``user.username`` and ``.username user`` are two acceptable
+way of accessing the username property of our user, we can simplify
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
ways of accessing the username property of our user, we can simplify
```

File: content/2019.04.getting-started-with-elm.rst [hunk 33]
```diff
@@ -0,0 +1,1100 @@
...
+our ``getUserName`` function like that:
+
+.. code-block:: elm
+
+    getUserName :  User -> String
+    getUserName =
+        .username
+
+It is really likely that the state of our app would grow, that's why
+it is recommended to use a type alias for ``Model``.
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]: Not sure I get the point here

File: content/2019.04.getting-started-with-elm.rst [hunk 34]
```diff
@@ -0,0 +1,1100 @@
...
+
+
+Elm types, a step further
+-------------------------
+
+Elm types can go a step further and we can create composite types, the
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
Elm types can go a step further, let's look at composite types. The
```

File: content/2019.04.getting-started-with-elm.rst [hunk 35]
```diff
@@ -0,0 +1,1100 @@
...
+``Maybe Int`` will then be either a ``Just value`` with value of
+type ``Int`` or a ``Nothing``.
+
+We have the same thing with ``Result`` that can be defined like that:
+
+.. code-block:: elm
+
+    type Result a b = Err a | Ok b
+
+We can then define ``Result Http.Error String`` that will be either
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
We can then define ``Result Http.Error String`` that is either
```

File: content/2019.04.getting-started-with-elm.rst [hunk 36]
```diff
@@ -0,0 +1,1100 @@
...
+``Ok value`` with ``value`` of type ``String`` or ``Err error`` with
+``error`` of type ``Http.Error``
+
+What can blow your mind is to have complex types that are self-explanatory:
```
Comment (coulix) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
What can blow your mind is to have complex types that are self-explanatory:
```

You mean in an impressed way or in a bad way?
reword.

**Comment by Natim** [comment]
Yannick [12 h 18]
Voila un exemple avec redux que j’ai gribouillé vite fait
Redux
```javascript
const initialState = 0;

const incrementAction = {
  type: "INCREMENT"
};

const decrementAction = {
  type: "DECREMENT"
};

function counterReducer(state = initialState, action) {
  switch (action.type) {
    case "INCREMENT":
      return state + 1;

    case "DECREMENT":
      return state - 1;

    default:
      return state;
  }
}
```

Elm

```elm
type alias Model = Int

type Action = 
     Increment 
   | Decrement

update : Action -> Model -> Model
update action model =
  case action of
    Increment ->
      model + 1

    Decrement ->
      model - 1
```

File: content/2019.04.getting-started-with-elm.rst (hunk 10 above)
Comment (Natim) [reply content/2019.04.getting-started-with-elm.rst]: Nope, Elm events

File: content/2019.04.getting-started-with-elm.rst (hunk 21 above)
Comment (Natim) [reply content/2019.04.getting-started-with-elm.rst]: It provides an API but I will rephrase thanks.

File: content/2019.04.getting-started-with-elm.rst [hunk 37]
```diff
@@ -0,0 +1,1100 @@
...
+
+For instance ``Maybe``, we can define maybe like that:
+
+.. code-block:: elm
+
+    type Maybe a = Just a | Nothing
+
+You might have seen the lowercase ``a`` here.
+
+It is just to tell elm that it can be any types. The only thing that
```
Comment (Natim) [inline content/2019.04.getting-started-with-elm.rst]:
```suggestion
It is just to tell Elm that it can be any types. The only thing that
```

**Pull request merged**